### PR TITLE
added With and Rt

### PR DESCRIPTION
**Pull request by Abdullahjavednesar**

### CODE REVIEW

**Comment by arihantparsoya** [comment]
Here are the list of functions that needs to be implemented:

`*` are the ones I am planning to implement:

- [ ] ExpandIntegrand
- [ ] * SumSimplerQ (Needs reimplementation)
- [ ] * SimplerQ (Needs reimplementation)
- [ ] TogetherSimplify
- [ ] Simp
- [x] * Defer (https://reference.wolfram.com/language/ref/Defer.html)

File: sympy/rubi/utility_function.py [hunk 1]
```diff
@@ -87,6 +87,14 @@ def RemoveContent(expr, x):
             expr = expr - i
     return expr
 
+def Rt(val, n):
+    return nthroot(val, n)
+
+def With(expr, **kwargs):
```
Comment (arihantparsoya) [inline sympy/rubi/utility_function.py]: I have already implemented `With`

Comment (Abdullahjavednesar) [reply sympy/rubi/utility_function.py]: Okay I'll remove this.

Comment (Abdullahjavednesar) [reply sympy/rubi/utility_function.py]: Did you get any information about IntLinearcQ()?

Comment (arihantparsoya) [reply sympy/rubi/utility_function.py]:
```mathematica
(* IntLinearcQ[a,b,c,d,m,n,x] returns True iff (a+b*x)^m*(c+d*x)^n is integrable wrt x in terms of non-hypergeometric functions. *)
IntLinearcQ[a_,b_,c_,d_,m_,n_,x_] :=
  IntegerQ[m] || IntegerQ[n] || IntegersQ[3*m,3*n] || IntegersQ[4*m,4*n] || IntegersQ[2*m,6*n] || IntegersQ[6*m,2*n] || IntegerQ[m+n]
```

**Comment by arihantparsoya** [comment]
It would be better that you mention the functions you are planning to implement in this thread(probably in your first comment) so we can prevent working on same function.

**Comment by Abdullahjavednesar** [comment]
```Simp(expr, x) ``` here what exactly the second argument x signify?

**Comment by arihantparsoya** [comment]
I am not sure, there is no description give in `Integration utility functions.m` file. Anyway, I think its better to implement the function similar to what they have done in mathematica. i.e. implementing `TimeConstrained`, `NormalizeSumFactors`, etc.

**Comment by Abdullahjavednesar** [comment]
@Upabjojr we need help for the remaining functions.

**Comment by arihantparsoya** [comment]
### List of functions that needs to be implemented: 

#### Set 1

- [x] *MapAnd
- [x] *Map2
- [ ] *ReapList
- [ ] *SetAttributes
- [x] *ComplexNumberQ
- [x] *RealNumericQ
- [x] *PositiveOrZeroQ
- [x] *NegativeOrZeroQ
- [x] *FractionOrNegativeQ
- [x] *SqrtNumberQ
- [x] SqrtNumberSumQ
- [x] IndependentQ
- [ ] NotIntegrableQ
- [x] PowerQ
- [x] ProductQ
- [x] SumQ
- [x] NonsumQ 
- [x] IntegerPowerQ
- [x] PositiveIntegerPowerQ
- [x] FractionalPowerQ

#### Set 2

- [x] *FractionalPowerFreeQ
- [x] *ComplexFreeQ
- [x] *ExpQ
- [x] *LogQ
- [x] *TrigQ
- [x] *SinQ, CosQ, TanQ, CotQ, SecQ, CscQ
- [x] *HyperbolicQ
- [x] *SinhQ, CoshQ, .....
- [x] InverseTrigQ
- [x] InverseHyperbolicQ
- [x] SinCosQ
- [x] SinhCoshQ
- [ ] CalculusQ
- [ ] CalculusFreeQ
- [ ] HeldFormQ

**Comment by Upabjojr** [comment]
For which ones do you need help?

**Comment by Abdullahjavednesar** [comment]
@Upabjojr, we are now implementing from the very first utility function as the later one's are dependent on them. Hopefully we'll implement the above list by tomorrow. If we need help we'll let you know.

**Comment by arihantparsoya** [comment]
I have implemented few functions. I am commenting out `def Int()` for now since it compiles Rubi rules which takes time. I will remove the comments when I test rubi-test-suit.

**Comment by arihantparsoya** [comment]
### Set 3 (Arihant)
- [x] InverseFunctionQ
- [x] TrigHyperbolicFreeQ
- [x] InverseFunctionFreeQ
- [x] ElementaryExpressionQ
- [x] RealQ
- [x] EqQ
- [x] PolyQ
- [x] BinomialQ
- [x] TrinomialQ
- [x] GeneralizedBinomialQ
- [x] GeneralizedTrinomialQ
- [x] PosQ
- [x] PosAux
- [x] NegQ
- [x] NiceSqrtQ
- [x] NiceSqrtAuxQ
- [x] PerfectSquareQ


### Set 4 (Abdullah)
- [x] SimplerQ
- [x] SimplerSqrtQ
- [ ] SumSimplerQ
- [ ] SumSimplerAuxQ
- [ ] SimplerIntegrandQ
- [ ] CancelCommonFactors
- [x] BinomialDegree
- [x] BinomialParts
- [x] TrinomialParts
- [ ] TrinomialDegree

File: sympy/rubi/tests/test_utility_function.py [hunk 2]
```diff
@@ -269,4 +269,4 @@ def test_PositiveIntegerPowerQ():
 
 def test_FractionalPowerQ():
     assert FractionalPowerQ(a**2) == True
-    assert FractionalPowerQ(a**sqrt(2)) == False
+    assert FractionalPowerQ(a**sqrt(2)) == False
```
Comment (arihantparsoya) [inline sympy/rubi/tests/test_utility_function.py]: complete the tests

File: sympy/rubi/utility_function.py [hunk 3]
```diff
@@ -289,3 +289,84 @@ def PositiveIntegerPowerQ(u):
...
+    return PowerQ(u) and u.args[1] == E
+
+def LogQ(u):
+    return expr.func == log
+
+def Head(u):
+    return u.func
+
+def MemberQ(l, u):
+    try:
```
Comment (arihantparsoya) [inline sympy/rubi/utility_function.py]: you can write `return u in l`

File: sympy/rubi/utility_function.py [hunk 4]
```diff
@@ -346,3 +289,84 @@ def PositiveIntegerPowerQ(u):
...
+        return any(expr == (u.func) for expr in l)
+    except TypeError:
+        return l == u.func
+
+def TrigQ(u):
+    if AtomQ(u):
+        x = u
+    else:
+        x = Head(u)
+return ([sin, cos, tan, cot, sec, csc], x)
```
Comment (Abdullahjavednesar) [inline sympy/rubi/utility_function.py]: it should be ```MemberQ([..``` I'll do that.

File: sympy/rubi/tests/test_utility_function.py [hunk 5]
```diff
@@ -270,3 +273,99 @@ def test_PositiveIntegerPowerQ():
 def test_FractionalPowerQ():
     assert FractionalPowerQ(a**2) == True
     assert FractionalPowerQ(a**sqrt(2)) == False
+
```
Comment (arihantparsoya) [inline sympy/rubi/tests/test_utility_function.py:276]: tests are passing right?

Comment (Abdullahjavednesar) [reply sympy/rubi/tests/test_utility_function.py:276]: yup

Comment (arihantparsoya) [reply sympy/rubi/tests/test_utility_function.py:276]: okay, some functions like `FractionalPowerFreeQ` are not implemented yet. You can implement them in next PR.

**Comment by arihantparsoya** [comment]
### Set 5(Abdullah)
- [x] GeneralizedBinomialDegree
- [x] GeneralizedBinomialParts
- [x] GeneralizedTrinomialParts
- [x] PowerOfLinearQ
- [x] QuadraticQ 
- [x] LinearPairQ
- [x] MonomialQ
- [x] MonomialSumQ
- [x] MinimumMonomialExponent 
- [x] MonomialExponent
- [x] LinearMatchQ


### Set 6(Arihant)
- [x] IntPart
- [x] FracPart
- [x] NumericFactor
- [x] NonnumericFactors
- [x] RemoveContent
- [x] RemoveContentAux
- [x] FreeFactors
- [x] NonfreeFactors
- [x] FreeTerms
- [x] NonfreeTerms
- [x] Expon
- [x] Coeff
- [x] LeadTerm
- [x] RemainingTerms
- [x] LeadFactor
- [x] LeadBase
- [x] LeadDegree
- [x] Numer
- [x] Denom

**Comment by arihantparsoya** [comment]
### Set 7 (Abdullah)
- [ ] PowerOfLinearMatchQ
- [ ] QuadraticMatchQ
- [x] CubicMatchQ
- [ ] BinomialMatchQ
- [ ] GeneralizedBinomialMatchQ
- [ ] TrinomialMatchQ
- [ ] GeneralizedTrinomialMatchQ
- [ ] QuotientOfLinearsMatchQ
- [ ] PolynomialTermQ
- [ ] PolynomialTerms
- [ ] NonpolynomialTerms
- [ ] PseudoBinomialQ
- [ ] PseudoBinomialPairQ
- [ ] NormalizePseudoBinomial
- [x] PseudoBinomialParts

### Set 8 (Arihant)
- [x] PerfectPowerTest
- [x] SquareFreeFactorTest
- [x] RationalFunctionQ
- [x] RationalFunctionFactors
- [x] NonrationalFunctionFactors
- [x] RationalFunctionExponents
- [x] RationalFunctionExpand
- [x] PolyGCD
- [x] AlgebraicFunctionQ
- [x] AlgebraicFunctionFactors
- [x] NonalgebraicFunctionFactors
- [x] QuotientOfLinearsQ
- [x] QuotientOfLinearsP
- [x] QuotientOfLinearsParts
- [x] SubstForFractionalPowerOfQuotientOfLinears
- [x] SubstForFractionalPowerQ
- [x] SubstForFractionalPowerAuxQ

**Comment by arihantparsoya** [comment]
#### Set 9 (Arihant)
- [x] AbsurdNumberQ
- [x] AbsurdNumberFactors
- [x] NonabsurdNumberFactors
- [x] FactorAbsurdNumber
- [x] CombineExponents
- [x] AbsurdNumberGCD
- [x] AbsurdNumberGCDList
- [x] FractionalPowerOfSquareQ
- [x] FractionalPowerSubexpressionQ
- [ ] FixSimplify
- [ ] SimpFixFactor
- [x] FactorNumericGcd
- [x] MergeFactors
- [x] MergeFactor
- [x] MergeableFactorQ
- [x] TrigSimplifyQ
- [x] TrigSimplify (sympy has `trigsimp` lets use that for now)

#### Set 10 (Abdullah)
- [ ] NormalizeIntegrand
- [ ] NormalizeIntegrandAux
- [ ] NormalizeIntegrandFactor
- [ ] NormalizeIntegrandFactorBase
- [ ] NormalizeTogether
- [ ] AbsorbMinusSign
- [ ] NormalizeSumFactors
- [ ] SignOfFactor
- [ ] NormalizePowerOfLinear
- [ ] MergeMonomials
- [ ] SimplifyIntegrand
- [ ] SimplifyTerm
- [ ] TogetherSimplify
- [ ] SmartSimplify
- [ ] SubstForExpn
- [ ] Simp
- [ ] SimpHelp

**Comment by arihantparsoya** [comment]
#### Set 11 (Arihant)
- [x] ContentFactor
- [x] UnifyNegativeBaseFactors(used by `ContentFactor` only hence no need to implement)
- [x] CommonFactors
- [x] MostMainFactorPosition
- [x] FactorOrder
- [x] Smallest
- [x] MonomialFactor
- [x] MinimumDegree
- [x] ConstantFactor
- [x] PositiveFactors
- [x] NonpositiveFactors
- [x] PolynomialInQ
- [x] ExponentIn
- [x] PolynomialInSubst
- [x] PolynomialDivide
- [x] Distrib
- [ ] Dist
- [x] DistributeDegree
- [x] FunctionOfPower
- [x] FunctionOfLinear
- [x] FunctionOfLinearSubst
- [x] DivideDegreesOfFactors

#### Set 12 (Abdullah)
- [x] ExpandToSum
- [ ] ExpandTrig
- [ ] ExpandExpression (already implemented using .expand())
- [ ] SmartApart
- [ ] MakeAssocList
- [ ] GensymSubst
- [ ] KernelSubst
- [ ] ExpandAlgebraicFunction
- [ ] UnifySum
- [ ] UnifyTerms
- [ ] ExpandTrigExpand
- [ ] ExpandTrigReduce
- [ ] ExpandTrigReduceAux
- [ ] ExpandTrigToExp

**Comment by arihantparsoya** [comment]
### Set 13(Abdullah)
- [ ] FunctionOfTrigOfLinearQ
- [ ] FunctionOfTrig
- [ ] FunctionOfTrig
- [ ] AlgebraicTrigFunctionQ
- [ ] FunctionOfHyperbolic
- [ ] FunctionOfQ
- [ ] FunctionOfExpnQ
- [ ] PureFunctionOfSinQ
- [ ] PureFunctionOfCosQ
- [ ] PureFunctionOfTanQ
- [ ] PureFunctionOfCotQ
- [ ] FunctionOfSinQ
- [ ] FunctionOfCosQ
- [ ] FunctionOfTanQ
- [ ] OddTrigPowerQ
- [ ] FunctionOfTanWeight
- [ ] FunctionOfTrigQ
- [ ] FunctionOfDensePolynomialsQ
- [ ] FunctionOfLog
- [ ] PowerVariableExpn
- [ ] PowerVariableDegree
- [ ] PowerVariableSubst
- [ ] EulerIntegrandQ
- [ ] FunctionOfSquareRootOfQuadratic
- [ ] SquareRootOfQuadraticSubst

### Set 14 (Arihant)
- [x] FunctionOfInverseLinear
- [x] FunctionOfExponentialQ
- [x] FunctionOfExponential
- [x] FunctionOfExponentialFunction
- [x] FunctionOfExponentialFunctionAux
- [x] FunctionOfExponentialTest
- [x] FunctionOfExponentialTestAux
- [x] PureFunctionOfSinhQ
- [x] PureFunctionOfCoshQ
- [x] PureFunctionOfTanhQ
- [x] PureFunctionOfCothQ
- [x] FunctionOfSinhQ
- [x] FunctionOfCoshQ
- [x] FunctionOfTanhQ
- [x] OddHyperbolicPowerQ
- [x] FunctionOfTanhWeight
- [x] FunctionOfHyperbolicQ
- [x] FindTrigFactor
- [ ] Subst
- [ ] SubstAux
- [ ] SimplifyAntiderivative
- [ ] SimplifyAntiderivativeSum
- [ ] RectifyTangent
- [ ] RectifyCotangent
- [x] SmartNumerator
- [ ] SubstFor
- [x] SubstForAux

**Comment by arihantparsoya** [comment]
### Set 15 (Arihant)

- [x] SubstForTrig
- [x] SubstForHyperbolic
- [x] SubstForFractionalPowerOfLinear
- [x] FractionalPowerOfLinear
- [x] InverseFunctionOfLinear
- [x] TryPureTanSubst
- [x] TryTanhSubst
- [x] InertTrigQ
- [x] InertReciprocalQ
- [x] InertTrigFreeQ
- [x] DeactivateTrig
- [x] DeactivateTrigAux
- [ ] FixInertTrigFunction
- [x] PowerOfInertTrigSumQ
- [x] KnownTangentIntegrandQ
- [x] PiecewiseLinearQ

### Set 16 (Abdullah)

- [ ] Divides
- [ ] DerivativeDivides
- [ ] EasyDQ
- [ ] ProductOfLinearPowersQ
- [ ] TrigSquare
- [ ] NthRoot
- [ ] RtAux
- [ ] AtomBaseQ
- [ ] SumBaseQ
- [ ] NegSumBaseQ
- [ ] AllNegTermQ
- [ ] SomeNegTermQ
- [ ] TrigSquareQ
- [ ] SplitProduct
- [ ] SplitSum
- [ ] IntSum
- [ ] IntTerm
- [ ] FixIntRules
- [ ] FixIntRule
- [ ] FixRhsIntRule

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , 2 tests are failing can you fix them?:
```python
____________________________________________________________________________________________________
_____________________ sympy/rubi/tests/test_utility_function.py:test_SimplerQ ______________________
  File "/users/parsoyaarihant/sympy/sympy/rubi/tests/test_utility_function.py", line 689, in test_SimplerQ
    assert not SimplerQ(2*x, x + 2)
AssertionError
____________________________________________________________________________________________________
____________________ sympy/rubi/tests/test_utility_function.py:test_expandTrig _____________________
  File "/users/parsoyaarihant/sympy/sympy/rubi/tests/test_utility_function.py", line 1209, in test_expandTrig
    assert ExpandTrig(sin(x) + 6, cos(x), x) == sin(x) + 6
AssertionError
```
If the error is in my part, then tell the function and expression, I will fix it.

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , there is a bug in `AbsorbMinusSign `:
Mathematica:
screen shot 2017-07-11 at 23 51 05
Python:
`assert AbsorbMinusSign((x + 2)**5*(x + 3)**5) == (-x - 3)**5*(-x - 2)**5`

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , bug in `NormalizeIntegrand`
Mathematica:
screen shot 2017-07-12 at 14 29 26
Python:
```python
>>> NormalizeIntegrand(b**2/(a**2*(a + b*x)**2), x)
1/(4*a**2*x**2)
```

Please try to fix these bugs as soon as possible.

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , `SimplifyTerm(a**2/b**2, x)` returns `nan`.

**Comment by Abdullahjavednesar** [comment]
@parsoyaarihant bug is in `MergeMonomials()` I guess, it returns `nan` for the same input.

**Comment by arihantparsoya** [comment]
alright, can you fix it?

**Comment by Abdullahjavednesar** [comment]
Ok, I'll do it.

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar ,  bug in `NormalizeIntegrand`
Mathematica:
screen shot 2017-07-13 at 21 55 53

Python:
`NormalizeIntegrand(6*q/(-16*q**7 + 16*q**5*x**2), x)` returns `nan`

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , tests for `SumSimplerQ` and `SumSimplerAuxQ` are failing in Python2.7. Can you fix them?

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , `QuadraticMatchQ`can also take `list` as arguments. Currently it does not support lists

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , can you implement `SplitProduct`, `SumBaseQ`?

**Comment by arihantparsoya** [comment]
`GeneralizedTrinomialQ` can take lists

**Comment by Abdullahjavednesar** [comment]
Okay I will do those.

**Comment by arihantparsoya** [comment]
Remaining Functions:

- [x] FixSimplify
- [x] SimpFixFactor
- [ ] Dist
- [ ] SubstFor
- [ ] SimplifyAntiderivative
- [ ] SimplifyAntiderivativeSum
- [ ] TrigSimplify (using `trigsimp` currently but that doesn't cover all the rules in utility functions)

**Comment by arihantparsoya** [comment]
@Abdullahjavednesar , add test for `Rt`

**Comment by Abdullahjavednesar** [comment]
@parsoyaarihant check [here](https://github.com/parsoyaarihant/sympy/blob/rubi4/sympy/integrals/rubi/tests/test_utility_function.py#L1615)

**Comment by arihantparsoya** [comment]
Okay, there was a bug in `RtAux()`. I have fixed it.

**Pull request merged**